transport/http/server.go

```diff
@@ -169,6 +171,45 @@ func EncodeJSONResponse(_ context.Context, w http.ResponseWriter, response inter
 	return json.NewEncoder(w).Encode(response)
 }
 
+// EncodeProtoResponse is an EncodeResponseFunc that serializes the response as Protobuf.
+// Many Proto-over-HTTP services can use it as a sensible default. If the response
+// implements Headerer, the provided headers will be applied to the response. If the
+// response implements StatusCoder, the provided StatusCode will be used instead of 200.
+func EncodeProtoResponse(ctx context.Context, w http.ResponseWriter, pres interface{}) error {
+	res, ok := pres.(proto.Message)
+	if !ok {
+		return errors.New("response does not implement proto.Message")
+	}
+	w.Header().Set("Content-Type", "application/x-protobuf")
+	if headerer, ok := w.(Headerer); ok {
+		for k := range headerer.Headers() {
+			w.Header().Set(k, headerer.Headers().Get(k))
+		}
+	}
+	code := http.StatusOK
+	if sc, ok := pres.(StatusCoder); ok {
+		code = sc.StatusCode()
+	}
+	w.WriteHeader(code)
+	if code == http.StatusNoContent {
+		return nil
+	}
+	if res == nil {
+		return nil
+	}
+	b, err := proto.Marshal(res)
+	if err != nil {
+		// returning an error here will cause a second header write, swallowing :(
```

Comment:
I think you have to return the error and trigger the second header write, otherwise there's no way clients are going to know this didn't work. Am I being naïve?

Reply:
Nah, I've been on the fence but leaning this way. After this nudge, we're returning the error.